### PR TITLE
Implement wrapping of EPM types

### LINKED ISSUE #413: Wrap EPM `Snapshot`, `SnapshotMetadata` and `DesiredTargets` to ensure invariants
The `Snapshot`, `SnapshotMetadata` and `DesiredTargets` storage items in the [EPM pallet](https://paritytech.github.io/substrate/master/pallet_election_provider_multi_phase/index.html) must be kept in sync throughout the election lifetime and be killed in tandem. Currently, these storage items are mutated in different places in the pallet. 

It makes sense to refactor EPM to hide those storage items behind a shared wrapper that is used to access/mutate and kill the inner items, so as to make sure the invariants hold true at every point in the election.

([related comment](https://github.com/paritytech/substrate/pull/13979#discussion_r1176946709))

### PR DESCRIPTION
**Pull request by wirednkod**
This PR wraps the `Snapshot`, `SnapshotMetadata` and `DesiredTargets` storage items in the [EPM pallet](https://paritytech.github.io/substrate/master/pallet_election_provider_multi_phase/index.html) in order to keep them in sync throughout the election lifetime and in order to be killed together. 

Prior to this PR, these storage items were mutated in different places in the pallet;

In addition 2 helper `fns` are introduced for chekcing if all the wrapped storage items exist or not;

Fixes #413 ;

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721279

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721280

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721281

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721436

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721437

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721435

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721843

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 3/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721844

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3721842

File: substrate/frame/election-provider-multi-phase/src/lib.rs [hunk 1]
```diff
@@ -1353,6 +1355,35 @@ pub mod pallet {
 	pub struct Pallet<T>(_);
 }
 
+pub struct SnapshotWrapper<T>(PhantomData<T>);
```
Comment (lexnv) [inline substrate/frame/election-provider-multi-phase/src/lib.rs:1358]: nit: Could add a small documentation that this wrapper handles the synchronization of `Snapshot`, `SnapshotMetadata` and `DesiredTargets`

**Review by lexnv — APPROVED** [review]
LGTM! Disclaimer: i'm not super familiar with the code base

**Review by sergejparity — APPROVED** [review]

**Review by kianenigma — CHANGES_REQUESTED** [review]
diff is broken. Blocking to make sure it is not merged before someone from @paritytech/staking-core approves.

**Comment by wirednkod** [comment]
Thank you @kianenigma . I messed up the repo configuration from merge to rebase. I will open a new PR fixing the mess. Apologies

**Comment by wirednkod** [comment]
Closing this in favour of #1633

**Pull request closed without merging**